backend/httpclient/options.go

```diff
@@ -0,0 +1,87 @@
+package httpclient
+
+import (
+	"time"
+)
+
+// Options defines options for creating HTTP clients.
+type Options struct {
+	// Timeouts timeout/connection related options.
+	Timeouts *TimeoutOptions
+
+	// BasicAuth basic authentication related options.
+	BasicAuth *BasicAuthOptions
+
+	// TLS TLS related options.
+	TLS   *TLSOptions
+	SigV4 *SigV4Config
+
+	// Headers custom headers.
+	Headers map[string]string
+
+	// CustomOptions allows custom options to be provided.
+	CustomOptions map[string]interface{}
+
+	// Labels could be used by certain middlewares.
+	Labels map[string]string
+
+	// Middlewares optionally provide additional middlewares.
+	Middlewares []Middleware
+
+	// ConfigureMiddleware optionally provide a ConfigureMiddlewareFunc
+	// to modify the middlewares chain.
+	ConfigureMiddleware ConfigureMiddlewareFunc
+}
+
+// BasicAuthOptions basic authentication options.
+type BasicAuthOptions struct {
+	User     string
+	Password string
+}
+
+// TimeoutOptions timeout/connection options.
+type TimeoutOptions struct {
+	Timeout               time.Duration
+	KeepAlive             time.Duration
+	TLSHandshakeTimeout   time.Duration
+	ExpectContinueTimeout time.Duration
+	MaxIdleConns          int
+	IdleConnTimeout       time.Duration
+}
+
+// DefaultTimeoutOptions default timeout/connection options.
+var DefaultTimeoutOptions = TimeoutOptions{
+	Timeout:               30 * time.Second,
+	KeepAlive:             30 * time.Second,
+	TLSHandshakeTimeout:   10 * time.Second,
+	ExpectContinueTimeout: 1 * time.Second,
+	MaxIdleConns:          100,
```

Comment:
I suggest increasing setting default `MaxIdleConnsPerHost` to a similar value, otherwise it's possible to quickly run out of available ports due to TIME_WAIT state of socket when sending requests fast enough. See https://tleyden.github.io/blog/2016/11/21/tuning-the-go-http-client-library-for-load-testing/

Reply:
It depends on if the client is used for multiple hosts, right? As explained in other comment the idea is to reuse a client per app/data source instance settings.

Reply:
This strikes if HTTP client makes frequent requests to the same host.

Reply:
UPD: frequent and concurrent (the concurrent is important since that's the reason why default max idle host 2 is not enough and results into many socket open/close)

Reply:
Good point didn't actually think it will do that :+1: